Add fields for ACME registration:

intermernet opened the pull request:
Added email address and TOS agreement fields.

These should be specified in the j8acfg.yml file in format:

connection:
  downstream:
    tls:
      acme:
        provider: letsencrypt
        email: noreply@example.org
        acceptTOS: true
        domains:
          - example.org

Probably needs extra tests, but currently works.

I've verified that it receives an error message from the ACME provider if the terms of service are not accepted,
but there is currently no validation of the email address. The email address is currently just passed to the ACME
provider for validation.

gofmt also re-ordered the imports in acmehandler.go. It's probably worth running gofmt across the entire code-base to prevent these changes polluting future PRs.

intermernet commented:
I'd like to add field validation and proper errors to `config.go` but I'd like some direction. Right now there are only 2 fields in the ACME config section, so it's feasible to do checks on all permutations:

https://github.com/simonmittag/j8a/blob/209141131bc1a3bfd8f627104da38b777b52b45f/config.go#L207-L266

In order to validate further fields it would be easier to nest the sub-checks within the original `if` statement that checks that any of the fields exists:

```go
if acmeProvider || acmeDomain || acmeEmail || acmeAcceptTOS{
	if len(config.Connection.Downstream.Tls.Cert) > 0 {
		config.panic("cannot specify TLS cert with ACME configuration, it would be overridden.")
	}

	if len(config.Connection.Downstream.Tls.Key) > 0 {
		config.panic("cannot specify TLS private key with ACME configuration, it would be overridden.")
	}

	if config.Connection.Downstream.Http.Port != 80 {
		config.panic("HTTP listener must be configured and set to port 80 for ACME challenge")
	}

	if !acmeProvider {
		config.panic("ACME provider must be specified with ACME config")
	}
	
	if !acmeDomain {
		config.panic("ACME domain must be specified with ACME config")
	}
	
	if !acmeEmail {
		config.panic("ACME email address must be specified with ACME config")
	}

	if !acmeAcceptTOS {
		config.panic("ACME TOS acceptance must be specified with ACME config")
	}

	if acmeDomain {
		.
		.
		.
	}
}
```

Any thoughts or feedback on this? I'm happy to put this together and add it to this PR.

simonmittag commented:
Hey Mike,

Thank you for the pull request, good idea to add the email! We probably need email address only in this case? What is the purpose of setting `acceptTos` to false? It will just panic the server. Are you attempting to get the user to "accept" the TOS (for which there is no link in the server) by placing a boolean: true in the config file explicitly? If yes, we'd need to include this in the documentation but I'm keen to get your feedback on why this flag is necessary in the first place? Alternatively, we could log.info() a link to the already accepted TOS? Happy to discuss.

> I'd like to add field validation and proper errors to `config.go` but I'd like some direction. Right now there are only 2 fields in the ACME config section, so it's feasible to do checks on all permutations:
> In order to validate further fields it would be easier to nest the sub-checks within the original `if` statement that checks that any of the fields exists:
> Any thoughts or feedback on this? I'm happy to put this together and add it to this PR.

1. I think the email address will need actual email format validation, as opposed to just a null check otherwise the registration may go wrong. Try to find a good package for this that is liberal with TLDs (I work for a company where email ends in @we.money)
2. The reason I test all permutations is that I never know which parameter is omitted. But your nesting looks fine to me. Tbh, the go code in config.validate is verbose and not pretty, but it's functional. Your nesting looks fine though.

If you're happy to add the email validation, let's have a chat how to fix the terms and then this can be merged.

simonmittag commented:
Also see: https://github.com/simonmittag/j8a/issues/264

intermernet commented:
Hi Simon,

I've just updated this PR with a bunch of extra tests and the modifications to the ACME config parsing. Have a look and see what you think.

Regarding the TOS, I was erring on the side of legal caution. I assume that accepting any TOS should be an explicit opt-in by the end user. If I remember correctly, caddy v1 had a command line flag to agree to the TOS, and also logged a message. I think caddy v2 asks for an email address if not specified and presents a message at that time saying that proceeding implies accepting the TOS.

I think it's reasonable to just log a message when ACME is specified saying something like "By using the ACME provider letsencrypt you agree with the terms and conditions of this provider. Please see their documentation for more details."

This can easily support additional ACME providers in the future, and I can even update the existing `acmeProviders` variable with friendly names and TOS URLs by creating an acmeProvider struct type with the details and changing `acmeProviders` from a `map[string]string` to a `map[string]acmeProvider`.

This would provide better logging, and more flexibility for the addition of future providers such as ZeroSSL.

intermernet commented:
Regarding parsing email addresses, I'm currently just using the `govalidator.IsEmail` function to check it. This looks like a fairly liberal regex test that should accept things like "simon@we.money". Email address parsing is a notoriously complex issue and I thought I'd just reuse a package that's already being imported rather than introduce more dependencies.

intermernet commented:
I've removed the `acceptTOS` field from the required config and instead logged a message of acceptance when an ACME provider is specified.

I've also re-factored the `ACMEproviders` variable and added an `ACMEprovider` type with additional fields for logging.

simonmittag commented:
Overall this looks good. I'm merging this to a branch on the local codebase to test it in the project's CI pipeline before release.

simonmittag commented:
Now running these changes at: https://adyntest.com/about We can release this with the next patch
